osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs

```diff
@@ -244,6 +244,8 @@ private void updateRepeats(IHasRepeats repeats)
 
         public override Vector2 ScreenSpaceSelectionPoint => ScreenSpaceDrawQuad.TopLeft;
 
+        protected override bool ComputeIsMaskedAway(RectangleF maskingBounds) => false;
```

Comment:
For pooled rulesets, this is fine. For non-pooled rulesets, I'm not sure it will be. I'm not sure how to test that, though, and I don't know if we care about that.

Explanation: Timeline blueprints are present on the timeline as long as their associated objects are alive (see: `HitObjectUsage{Began,Finished}`, `{Past,Future}LifetimeExtension`). For pooled paths this is nice and all, since there the entire lifetime start/end/hitobject entry machinery is working. But I don't know what happens for rulesets that don't do pooling and don't have any of that working. If it's as bad as I worry it is, then disabling masking entirely for all blueprints of a beatmap will destroy performance for non-pooled rulesets.

For my peace of mind, is there a way to fix this which expands the masking quad just enough for this issue to go away, but doesn't turn off masking entirely?

Reply:
I'm disabling masking only for this parent container, children are computing their own mask anyway. But yes, mask computation will be increased by children count, though I can't assure you that performance difference will be minuscule.
Since pills above and below have different sizes depending on their content (text length) - guessing quad expansion is a bit trickier.

Reply:
However we can use only object's quad as a mask, but in this case long pills are still affected
![update](https://github.com/ppy/osu/assets/22874522/ef469810-e103-4c29-bdba-3118bf976eab)
Still looks better than master

Reply:
How about

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs b/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs
index ea063e9216..51de074396 100644
--- a/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs
@@ -57,6 +57,13 @@ public partial class TimelineHitObjectBlueprint : SelectionBlueprint<HitObject>
         [Resolved]
         private OverlayColourProvider colourProvider { get; set; } = null!;
 
+        protected override bool ComputeIsMaskedAway(RectangleF maskingBounds)
+        {
+            var rect = RectangleF.Union(ScreenSpaceDrawQuad.AABBFloat, circle.ScreenSpaceDrawQuad.AABBFloat);
+
+            return !Precision.AlmostIntersects(maskingBounds, rect);
+        }
+
         public TimelineHitObjectBlueprint(HitObject item)
             : base(item)
         {
```

Has the same issue with the pills, but could be resolved with two more unions.

Reply:
Applied this logic and added some comments.